sycl/doc/extensions/GroupMask/GroupMask.asciidoc

```diff
@@ -0,0 +1,340 @@
+must be encountered by all work-items in the group in converged control flow.
+
+|===
+|Function|Description
+
+|`template <typename Group> Group::mask_type group_ballot(Group g, bool predicate = true) const`
+|Return a `group_mask` representing the set of work-items in group _g_ for which _predicate_ is `true`.
+|===
+
+=== Group Masks
+
+The group mask type is an opaque type, permitting implementations to use any
+mask representation that has the same size and alignment across host and
+device. The maximum number of bits that can be stored in a `group_mask` is
+exposed as a static member variable, `group_mask::max_bits`.
+
+Functions declared in the `group_mask` class can be called independently by
+different work-items in the same group.  An instance of a group class (e.g.
+`group` or `sub_group`) is not required to manipulate a group mask.
+
+The mask is defined such that the least significant bit (LSB) corresponds to
+the work-item with id 0, and the most significant bit (MSB) corresponds to the
+work-item with the id `max_local_range()-1`.
+
+|===
+|Member Function|Description
+
+|`bool operator[](id<1> id) const`
+|Return `true` if the bit corresponding to the specified _id_ is set in the
+ mask.
+
+|`group_mask::reference operator[](id<1> id)`
+|Return a reference to the bit corresponding to the specified _id_ in the mask.
+
+|`bool test(id<1> id) const`
+|Return `true` if the bit corresponding to the specified _id_ is set in the
+ mask.
+
+|`bool all() const`
+|Return `true` if all bits in the mask are set.
+
+|`bool any() const`
+|Return `true` if any bits in the mask are set.
+
+|`bool none() const`
+|Return `true` if none of the bits in the mask are set.
+
+|`uint32_t count() const`
+|Return the number of bits set in the mask.
+
+|`uint32_t size() const`
+|Return the number of bits in the mask.
+
+|`id<1> find_low() const`
+|Return the lowest `id` with a corresponding bit set in the mask. If no bits
+ are set, the return value is equal to `size()`.
+
+|`id<1> find_high() const`
+|Return the highest `id` with a corresponding bit set in the mask. If no bits
+ are set, the return value is equal to `size()`.
+
+|`template <typename T = marray<uint32_t, max_bits/sizeof(uint32_t)>> void insert_bits(T bits, id<1> pos = 0)`
+|Insert `CHAR_BIT * sizeof(T)` bits into the mask, starting from _pos_.  `T`
+ must be an integral type or a SYCL `marray` of integral types.  _pos_ must be a
+ multiple of `CHAR_BIT * sizeof(T)` in the range [0, `size()`).  If _pos_ pass:[+]
+ `CHAR_BIT * sizeof(T)` is greater than `size()`, the final `size()` - (_pos_ pass:[+]
+ `CHAR_BIT * sizeof(T)`) bits are ignored.
+
+|`template <typename T = marray<uint32_t, max_bits/sizeof(uint32_t)>> T extract_bits(id<1> pos = 0) const`
+|Return `CHAR_BIT * sizeof(T)` bits from the mask, starting from _pos_.  `T`
+ must be an integral type or a SYCL `marray` of integral types.  _pos_ must be a
+ multiple of `CHAR_BIT * sizeof(T)` in the range [0, `size()`).  If _pos_ pass:[+]
```

Comment:
> _pos_ must be a multiple of `CHAR_BIT * sizeof(T)`

Is this strictly true?  Consider the case where `T` is `marray<uint32_t, 4>`, for example.  In that case, `CHAR_BITS * sizeof(T)` will be 128.  Do we really want to restrict `pos` to be a multiple of 128?  Or, do we intend to say that `pos` must be a multiple of `CHAR_BITS * sizeof(uint32_t) = 32` in this case?

Reply:
Hm.  I don't think it _needs_ to be true... but it might be simpler, and possibly faster.

If `T` is a type like `char`, `uint32_t` or `uint64_t`, I think it makes sense for `pos` to be a multiple of the type itself.  If we do something different for `marray<T, N>`, to say that `pos` becomes a multiple of `T` instead, that might be surprising.  I think we settled on this design because it allows the behavior of `pos` to be consistent for all `T`.  We can't have the units of `pos` be the same for all `T` (which would have been the other option) because we can't allow anything less than a byte, and misaligned inserts/extracts might require a memcpy.

Reply:
OK, I just wanted to make sure that we wrote what we intended to say.